lib/core/src/Cardano/Wallet/Transaction.hs

```diff
@@ -187,20 +236,103 @@ withdrawalToCoin = \case
     WithdrawalExternal _ _ c -> c
     NoWithdrawal -> Coin 0
 
+withdrawalRewardAccount :: Withdrawal -> Maybe RewardAccount
+withdrawalRewardAccount = \case
+    WithdrawalSelf acct _ _ -> Just acct
+    WithdrawalExternal acct _ _ -> Just acct
+    NoWithdrawal -> Nothing
+
 -- | A default context with sensible placeholder. Can be used to reduce
 -- repetition for changing only sub-part of the default context.
 defaultTransactionCtx :: TransactionCtx
 defaultTransactionCtx = TransactionCtx
     { txWithdrawal = NoWithdrawal
     , txMetadata = Nothing
     , txTimeToLive = maxBound
-    , txDelegationAction = Nothing
+    , txDelegationActions = mempty
     }
 
 -- | Whether the user is attempting any particular delegation action.
-data DelegationAction = RegisterKeyAndJoin PoolId | Join PoolId | Quit
+data DelegationAction = RegisterKey | Join PoolId | Quit
     deriving (Show, Eq, Generic)
 
+data SignTransactionResult sk wit tx = SignTransactionResult
+    { tx :: !tx
+    , addressWitnesses :: ![SignTransactionWitness sk wit]
+    , withdrawalWitnesses :: ![(RewardAccount, wit)]
+    , certificateWitnesses :: ![(RewardAccount, wit)]
+    } deriving (Show, Eq, Generic, Functor)
+
+instance Bifunctor (SignTransactionResult sk) where
+    first f (SignTransactionResult t a w c) =
+        SignTransactionResult t (fmap (fmap f) a) (fmap (fmap f) w) (fmap (fmap f) c)
+    second f (SignTransactionResult t a w c) =
+        SignTransactionResult (f t) a w c
+
+-- | In the wallet, signing keys are symmetrically encrypted at rest using a key
+-- derived from the user's spending passphrasee. This data type pairs an
+-- encrypted signing key with the passphrase needed to decrypt it.
+--
+-- We call this type "decrypted", because although the key is still encrypted, the passphrase is right there, so it may as well be decrypted.
```

Comment:
```suggestion
-- We call this type "decrypted", because although the key is still encrypted,
-- the passphrase is right there, so it may as well be decrypted.
```